policyengine_us_data/datasets/cps/cps.py

```diff
@@ -838,42 +838,311 @@ def add_previous_year_income(self, cps: h5py.File) -> None:
+    # In Jan 2024, the total U.S. civilian labor forceis reported as 167.1 million people.
+    # https://fred.stlouisfed.org/series/CLF16OV
+    # Unauthorized immigrant workers is 8.3 million.
+    # https://www.pewresearch.org/short-reads/2024/07/22/what-we-know-about-unauthorized-immigrants-living-in-the-us/
+    # share of undocumented immigrant workers who are unauthorized to work is: 8.3 / (0.192 * 167.1)
+    worker_ids = person[worker_mask].index
+    n_worker_ead = int(0.74 * len(worker_ids))
+    selected_workers = np.random.choice(
+        worker_ids, size=n_worker_ead, replace=False
+    )
+
+    # undocumented immigrant students who account for approximately 21 percent of the total 1.9 million immigrant students
+    # https://www.higheredimmigrationportal.org/research/immigrant-origin-students-in-u-s-higher-education-updated-august-2024/
+    student_ids = person[student_mask].index
+    n_student_ead = int(0.79 * len(student_ids))
+    selected_students = np.random.choice(
+        student_ids, size=n_student_ead, replace=False
+    )
+
+    # Assign code 2
+    ssn_card_type[selected_workers] = 2
+    ssn_card_type[selected_students] = 2
+
+    final_counts = pd.Series(ssn_card_type).value_counts().sort_index()
+
+    # ============================================================================
+    # FAMILY CORRELATION ADJUSTMENT
+    # ============================================================================
+
+    # Identify parent-child relationships using household and family data
+    correlation_probability = 0.8
+    # Only applies to families with codes 0 or 3 (not citizens or valid EAD holders)
+    rng_family = np.random.default_rng(seed=123)
+
+    # Create a DataFrame for easier family processing
+    family_df = pd.DataFrame(
+        {
+            "person_id": person.PH_SEQ * 100 + person.P_SEQ,
+            "household_id": person.PH_SEQ,
+            "family_id": person.PH_SEQ * 10 + person.PF_SEQ,
+            "age": person.A_AGE,
+            "parent1_line": person.PEPAR1,  # Line number of first parent
+            "parent2_line": person.PEPAR2,  # Line number of second parent
+            "line_number": person.A_LINENO,
+            "ssn_code": ssn_card_type,
+        }
+    )
+
+    # Identify children (those with parent pointers)
+    children = family_df[
+        (family_df.parent1_line > 0) | (family_df.parent2_line > 0)
+    ]
 
-    # Code 2: Noncitizens (PRCITSHP == 5) who are working or studying
-    noncitizen_mask = person.PRCITSHP == 5
-    is_worker = (person.WSAL_VAL > 0) | (person.SEMP_VAL > 0)  # worker
-    is_student = person.A_HSCOL == 2  # student
-    ead_like_mask = noncitizen_mask & (is_worker | is_student)
-    ssn_card_type[ead_like_mask] = 2
+    families_adjusted = 0
 
-    # Step 3: Refine remaining 0s into 0 or 3
-    share_code_3 = 0.3  # IRS/SSA target share of SSA-benefit-only cards
+    for _, child in children.iterrows():
```

Comment:
could you explain how this works? are you overriding any flags from earlier in the logic?